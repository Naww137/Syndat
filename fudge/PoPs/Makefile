# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

SUB_DIRS = quantities chemicalElements families decays atomic Test

.PHONY: clean realclean doTarget

default:

clean:
	rm -f *.pyc
	$(MAKE) doTarget MODE=clean

realclean: clean
	$(MAKE) doTarget MODE=realclean

doTarget:
	SAVED_PWD=`pwd`; \
	for directory in $(SUB_DIRS); do cd $$directory; $(MAKE) $(MODE); cd $$SAVED_PWD; done
