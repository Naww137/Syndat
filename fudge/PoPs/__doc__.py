# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This is the 'Properties of Particles' or PoPs package. It defines a set of Python classes
for storing intrinsic particle properties like charge, mass, spin, etc.
"""
