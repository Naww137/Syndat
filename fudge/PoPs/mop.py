# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

__doc__ = """
This module contains classes to handle a MoP (Map of PoPs) instance.
"""

import os
import re

from LUPY import ancestry as ancestryModule
from . import database as databaseModule

class FormatVersion :

    version_0_1 = '0.1'

    allowed = ( version_0_1, )

class Mop(ancestryModule.AncestryIO):

    moniker = 'mop'

    def __init__(self, name, path = './', format=FormatVersion.version_0_1):

        """
        Constructor for a Mop instance.

        :param name:            The name of the library.
        :param path:            The path to the mop file if read in from disk.
        :param format:          The format version for the mop.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( name, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__name = name 

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__path = path

        if( format not in FormatVersion.allowed ) : raise ValueError( 'Invalid format "%s."' % format )
        self.__format = format

        self.__entries = []
        self.__pops = databaseModule.Database( 'Internal', '0.1' )

    def __len__( self ) :
        """Returns the number of entries in self."""

        return( len( self__entries ) )

    def __getitem__( self, index ) :
        """
        Returns the (index-1)^th item of self.

        :param index:       The index of the iter to return.
        """

        return( self.__entries[index] )

    @property
    def format( self ) :
        """Returns the format self is stored as."""

        return( self.__format )

    @property
    def name( self ) :
        """Returns self's name."""

        return( self.__name )

    @property
    def path( self ) :
        """Returns the path to self. If the data in self was read from disk, this is the path to the file."""

        return( self.__path )

    @property
    def fileName( self ) :
        """Same the path method."""

        return( self.__path )

    def append( self, entry ) :
        """Appends *entry* to the end of self's list of entries."""

        if( not( isinstance( entry, EntryBase ) ) ) : raise TypeError( 'Invalid instance to add.' )

        self.__entries.append( entry )
        entry.setAncestor( self )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns a list of items. Each item contains information about a PoPs database where one or more matches to *re_pid* exists. 
        Each item contains the list [ ids, all_ids, fileName, pops ]) where:

            -) ids:         The list of PoPs ids in a PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The file path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]).  *re_pid* can be a str instance or a *re.compile* instance.
        Note, even if *readPoPs* is False, the returned pops will not be None if it had previously been read in.
        
        :param re_pid:          A str or a *re.compile* pattern representing the ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        matches = []
        for entry in self.__entries : matches += entry.fetch( re_pid, readPoPs )

        return( matches )

    def toXML_strList( self, indent = "", **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        indent2 = indent + kwargs.get( 'incrementalIndent', '  ' )

        format = kwargs.get( 'format', FormatVersion.version_0_1 )
        if( format not in FormatVersion.allowed ) : raise TypeError( 'Invalid format = "%s".' % format )

        XML_list = [ '%s<%s name="%s" format="%s">' % ( indent, self.moniker, self.name, format ) ]
        for entry in self.__entries : XML_list += entry.toXML_strList( indent2, **kwargs )
        XML_list[-1] +=  '</%s>' % self.moniker

        return( XML_list )

    @staticmethod
    def parseNode( node, path ) :
        """
        Creates a Mop instance from an XML mop element.

        :param node:        XML element to parse.

        :return:            Mop instance.
        """

        if( node.tag != Mop.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Mop.moniker, node.tag ) )

        format = node.get( 'format', None )
        if( format is None ) : raise ValueError( 'Mop node does not have a "format" attribute.' )

        name = node.get( 'name', None )
        if( name is None ) : raise ValueError( 'Mop node does not have "name" attribute.' )

        mop = Mop( name, path = path )

        settings = { 'format' : format }
        for child in node :
            if( child.tag == Import.moniker ) :
                mop.append( Import.parseNode( settings, child ) )
            elif( child.tag == PoPs.moniker ) :
                mop.append( PoPs.parseNode( settings, child ) )
            else :
                raise ValueError( 'Invalid child tag "%s" for mop file.' % child.tag )

        return( mop )

class EntryBase(ancestryModule.AncestryIO) :
    """
    Base class for mop entries.
    """

    def __init__( self, path ) :
        """
        Base constructor for mop entries.

        :param path:            Path to the file for the entry.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Path must be a string.' )
        self.__path = path

    @property
    def fileName( self ) :
        """Returns the absolute path to the file pointed to by the 'path' attribute."""

        filename = self.path
        if( filename != os.sep ) : filename = os.path.join( os.path.dirname( os.path.realpath( self.ancestor.fileName ) ), filename )
        return( filename )

    @property
    def path( self ) :
        """Returns the path of the mop file. This may be absolute or relative."""

        return( self.__path )

class PoPs(EntryBase):
    """
    Class for storing a GNDS mop's PoPs instance.
    """

    moniker = 'pops'

    def __init__( self, path, ids ) :
        """
        Constructor for a mop's PoPs instance.

        :param path:            Path to the GNDS PoPs.database file.
        :param ids:             The list of particles ids in the PoPs.database file.
        """

        EntryBase.__init__(self, path)
        self.__pops = None
        self.__ids = ids

    @property
    def ids( self ) :
        """Returns the list of ids."""

        return( self.__ids )

    @property
    def pops( self ) :
        """Returns a reference to the PoPs.database instance. If the PoPs file has not be read in (see method *readIfNeeded*), None is returned."""

        return( self.__pops )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns the list [ ids, all_ids, fileName, pops ] if matches to *re_pid* are found in *self* and an empty list otherwise. 

            -) ids:         The list of PoPs ids in the PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]). *re_pid* can be a str instance or a *re.compile* instance.
        
        :param re_pid:          A str or *re.compile* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element.
        """

        if( isinstance( re_pid, str ) ) : re_pid = re.compile( re_pid )
        matches = [ pid for pid in self.ids if re_pid.fullmatch( pid ) ]

        if( len( matches ) > 0 ) :
            matches = [ [ matches, self.ids, self.fileName, self.readIfNeeded( readPoPs ) ] ]

        return( matches )

    def readIfNeeded( self, readPoPs = True ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in by a prior call and *readPoPs* 
        is True (default). Also, returns the value of *self.__pops*. The returned value will only be *None* if *readPoPs* 
        is *False* and all prior calls to this method had "*readPoPs* = False".

        :param readPoPs:        If *True*, the pops file is read.
        :return:                The value if *self.__pops*.
        """

        if( ( self.__pops is None ) and readPoPs ) :
            self.__pops = databaseModule.read( self.fileName )
            self.__pops.setAncestor( self )

        return( self.__pops )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s">' % ( indent, self.moniker, self.path ) + ' '.join( self.__ids ) + '</%s>' % self.moniker ] )

    @classmethod
    def parseNodeUsingClass(settings, node):
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != PoPs.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( PoPs.moniker, node.tag ) )

        path = node.get( 'path', None )
        ids = node.text.split( )

        pops = PoPs( path, ids )

        return( pops )

class Import(EntryBase):
    """
    Class for storing a GNDS mop's Import instance.
    """

    moniker = 'import'

    def __init__( self, path ) :
        """
        Constructor for a mop's import instance.

        :param path:            Path to the GNDS mop file.
        """

        EntryBase.__init__(self, path)
        self.__mop = None

    @property
    def mop( self ) :
        """Returns a reference to self.__mop."""

        return( self.__mop )

    def fetch( self, re_pid, readPoPs ) :
        """
        See documentation for Mop.fetch.
        
        :param re_pid:          A str or *re* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        self.readIfNeeded( )
        return( self.__mop.fetch( re_pid, readPoPs ) )

    def readIfNeeded( self ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in.
        """

        if( self.__mop is None ) :
            self.__mop = Mop.readXML_file( self.fileName )
            self.__mop.setAncestor( self )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s"/>' % ( indent, self.moniker, self.path ) ] )

    @staticmethod
    def parseNode( settings, node ) :
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != Import.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Import.moniker, node.tag ) )

        path = node.get( 'path', None )

        _import = Import( path )

        return( _import )
