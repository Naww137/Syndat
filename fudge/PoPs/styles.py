# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

import abc
import datetime

from LUPY import ancestry as ancestryModule
from xData.Documentation import documentation as documentationModule

class Styles(ancestryModule.AncestryIO_base):
    """
    Stores the list of PoPs styles that appear inside a database.
    """
    # FIXME merge with fudge.styles class?

    moniker = 'styles'

    def __init__( self ) :

        ancestryModule.AncestryIO.__init__(self)
        self.__styles = []

    def __contains__( self, label ) :

        for item in self :
            if( item.label == label ) : return( True )
        return( False )

    def __len__( self ) :

        return( len( self.__styles ) )

    def __iter__( self ) :

        n1 = len( self )
        for i1 in range( n1 ) : yield self.__styles[i1]

    def __getitem__( self, label ) :

        if( isinstance( label, int ) ) : return( self.__styles[label] )
        if( not( isinstance( label, str ) ) ) : raise TypeError( 'label must be a string' )
        for _style in self.__styles :
            if( _style.label == label ) : return( _style )
        raise IndexError( 'No style labelled == "%s"' % label )

    def add( self, _style ) :
        """
        Append a style to the list of styles.

        :param _style: style instance
        """

        if( not( isinstance( _style, Style ) ) ) : raise TypeError( 'invalid style instance' )

        for __style in self :
            if( __style.label == _style.label ) : raise ValueError( 'style labeled "%s" already exists' % _style.label )

        self.__styles.append( _style )
        _style.setAncestor(self)

    def convertUnits( self, unitMap ) :
        """See database.convertUnits"""

        for _style in self : _style.convertUnits( unitMap )

    def remove( self, _style ):
        """
        Remove the specified style. Raises KeyError if specified style not found.
        :param _style: may be a string or a style instance
        """

        index = None
        for idx,tmp in enumerate(self):
            if( ( tmp is _style ) or ( tmp.label == _style ) ) : index = idx
        if index is None:
            raise KeyError("style '%s' not found in styles" % _style)
        self.__styles.pop(index)

    def getEvaluatedStyle(self):
        _evaluateds = [_style for _style in self if isinstance(_style, Evaluated)]
        if len(_evaluateds) == 0:
            raise Exception("No evaluated style found.")
        if len(_evaluateds) > 1:
            raise Exception("Multiple (%d) evaluated styles found." % len(_evaluateds))
        return _evaluateds[0]

    def toXML_strList( self, indent = '', **kwargs ) :

        indent2 = indent + kwargs.get( 'incrementalIndent', '  ' )

        if( len( self ) == 0 ) : return( [] )

        xmlStringList = ['%s<%s>' % (indent, self.moniker)]
        for _style in self : xmlStringList += _style.toXML_strList( indent2, **kwargs )
        xmlStringList[-1] += '</%s>' % self.moniker
        return( xmlStringList )

    def parseNode(self, node, xPath, linkData, **kwargs):

        xPath.append(node.tag)

        classDict = {}
        for _style in ( Evaluated, ) : classDict[_style.moniker] = _style
        for child in node:
            _class = classDict.get( child.tag, None )
            if( _class is None ) :
                raise TypeError( 'encountered unknown style "%s"' % child.tag )
            self.add(_class.parseNodeUsingClass(child, xPath, linkData, **kwargs))


        xPath.pop()

class Style(ancestryModule.AncestryIO, abc.ABC):
    """ Abstract base class for all 'style' classes in PoPs """

    def __init__( self, label, derivedFrom, date = None ) :
        """
        :param label: string identifying this style
        :param derivedFrom: may be a string (equal to another style's label), or None if this is an original evaluation
        :param date: optional date when style was generated. Defaults to today's date
        """

        ancestryModule.AncestryIO.__init__( self )

        if( not( isinstance( label, str ) ) ) : raise TypeError( 'label must be a str instance.' )
        self.__label = label

        if( date is None ) : date = str( datetime.date.today( ) )
        self.__date = date

        if( not( isinstance( derivedFrom, str ) ) ) : raise TypeError( 'label must be a str instance.' )
        self.__derivedFrom = derivedFrom

        self.__documentation = documentationModule.Documentation()
        self.__documentation.setAncestor(self)

    @property
    def date( self ) :
        """Returns date when style was generated"""

        return( self.__date )

    @property
    def derivedFrom( self ) :
        """Returns string label of the style that self derives from"""

        return( self.__derivedFrom )

    @property
    def derivedFromStyle( self ) :
        """Returns style instance that self derives from"""

        for _style in self.ancestor :
            if( _style.label == self.__derivedFrom ) : return( _style )
        return( None )

    @property
    def documentation(self):
        return self.__documentation

    @property
    def label( self ) :

        return( self.__label )

    def findFormMatchingDerivedStyle( self, component, styleFilter = None ) :
        """
        This method searches backwards through the chain of derived styles to find a matching form in
        the specified component. It starts with the style that self derives from, then the style that derives from, etc.
        The first form matching one of those styles is returned (unless styleFilter is used to screen out some styles).

        :param component: a PoPs component, e.g. 'mass', 'spin' or 'energy'.
        :param styleFilter: optional function that can be used to filter out some matches. styleFilter should
            take a style as input and return False if that style should be omitted from matches
        :return: matching form, or None if no match is found
        """

        def alwaysTrue( a_style ) : return( True )

        if( styleFilter is None ) : styleFilter = alwaysTrue

        parent = self.ancestor
        derivedFrom = self.derivedFrom
        while( derivedFrom != '' ) :
            for form in component :
                if( ( derivedFrom == form.label ) and styleFilter( parent[derivedFrom] ) ) : return( form )
            derivedFrom = parent[derivedFrom].derivedFrom
        return( None )
        
    def findDerivedFromStyle( self, cls ) :
        """
        This method searches backwards through the chain of derived styles to find an instance of the specified
        class (which should inherit from the base style class).

        :param cls: desired class
        :return: matching class instance, or None if no match found
        """

        _style = self
        while( _style is not None ) :
            _style = _style.derivedFromStyle
            if( ( _style is not None ) and ( isinstance( _style, cls ) ) ) : return( _style )
        return( None )

    def sibling( self, label ) :
        """Returns the sibling of self with label."""

        return( self.ancestor[label] )

    def convertUnits( self, unitMap ) :

        pass

    def XMLCommonAttributes( self ) :
        """Helper function for writing to XML"""

        XMLCommon = 'label="%s"' % self.label
        if( self.derivedFrom != "" ) : XMLCommon += ' derivedFrom="%s"' % self.derivedFrom
        if( self.date is not None ) : XMLCommon += ' date="%s"' % self.date
        return( XMLCommon )

    @staticmethod
    def parseBaseNodeCommons(node, xPath, **kwargs):
        """
        Parse *label*, *derivedFrom* and *date* that are common to all style nodes. Also, set xPath.
        """

        label = node.get('label')
        xPath.append('%s[@label="%s"]' % (node.tag, label))

        derivedFrom = node.get('derivedFrom', '')
        date = node.get('date', None)

        return label, derivedFrom, date

class Evaluated(Style):

    moniker = 'evaluated'

    def __init__( self, label, derivedFrom, library = '', version = '', date = None ) :
        """
        :param label: a unique string identifying this style
        :param derivedFrom: may be a string (equal to another style's label), or None if this is an original evaluation
        :param library: optional string identifying the library, e.g. 'ENSDF'
        :param version: optional string identifying the library version
        :param date: optional date when style was generated. Defaults to today's date
        """

        Style.__init__( self, label, derivedFrom, date = date )

        if( not( isinstance( library, str ) ) ) : raise TypeError( 'library must be a string' )
        self.__library = library

        if( not( isinstance( version, str ) ) ) : raise TypeError( 'version must be a string' )
        self.__version = version

    @property
    def library( self ) :

        return( self.__library )

    @library.setter
    def library( self, value ) :

        self.__library = value

    @property
    def version( self ) :

        return( self.__version )

    @version.setter
    def version( self, value ) :

        self.__version = value

    def copy( self ) :

        return( Evaluated( self.label, self.derivedFrom, library = self.library, version = self.version, date = self.date ) )

    def toXML_strList( self, indent = '', **kwargs ) :

        indent2 = indent + kwargs.get('incrementalIndent', '  ')

        xmlStringList = [ '%s<%s %s library="%s" version="%s">' %
                ( indent, self.moniker, self.XMLCommonAttributes( ), self.library, self.version ) ]

        if kwargs.get('formatVersion') != '0.1':  # PoPs version corresponding to GNDS 1.10
            xmlStringList += self.documentation.toXML_strList( indent2, **kwargs )

        xmlStringList[-1] += '</%s>' % self.moniker
        return( xmlStringList )

    @classmethod
    def parseNodeUsingClass(cls, node, xPath, linkData, **kwargs):

        label, derivedFrom, date = Style.parseBaseNodeCommons(node, xPath, **kwargs)        # Also adds to xPath.

        library = node.get('library', '')
        version = node.get('version', '')

        _evaluated = cls( label, derivedFrom, library, version, date = date )

        _documentation = node.find(documentationModule.Documentation.moniker)
        if _documentation is not None:
            _evaluated.documentation.parseNode(_documentation, xPath, linkData, **kwargs)

        xPath.pop()                 # Need to pop as parseBaseNodeCommons added to it.

        return _evaluated
