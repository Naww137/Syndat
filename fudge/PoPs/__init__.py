# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
Top-level __init__ for the PoPs package.
"""

POPS_MAJORVERSION = 0
POPS_MINORVERSION = 1
POPS_PATCHLEVEL = 0
__version__ = '%s.%s.%s' % ( POPS_MAJORVERSION, POPS_MINORVERSION, POPS_PATCHLEVEL )

from .__doc__ import __doc__
