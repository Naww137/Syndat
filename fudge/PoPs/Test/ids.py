#! /usr/bin/env python3

# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

from PoPs import misc as miscModule

parts = miscModule.baseAntiQualifierFromID( 'e-' )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'e-_anti' )
print( parts )

parts = miscModule.baseAntiQualifierFromID( 'O' )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'O_anti' )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'O{abc}', qualifierAllowed = True )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'O_anti{abc}', qualifierAllowed = True )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'O16' )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'O16_anti' )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'O16{abc}', qualifierAllowed = True )
print( parts )
parts = miscModule.baseAntiQualifierFromID( 'O16_anti{abc}', qualifierAllowed = True )
print( parts )
