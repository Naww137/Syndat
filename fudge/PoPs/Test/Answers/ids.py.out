('e-', '', '')
('e-', '_anti', '')
('O', '', '')
('O', '_anti', '')
('O', '', 'abc')
('O', '_anti', 'abc')
('O16', '', '')
('O16', '_anti', '')
('O16', '', 'abc')
('O16', '_anti', 'abc')
